### PR TITLE
Knot Engine implementation

### PR DESCRIPTION
**Pull request by tomaszmichalak**
Knot Engine implementation with Unit tests.

## Description
Knot's contract has changed. 
In Knot.x 1.0, Knot worked on all fragments.
In Knot.x 2.0, Knot works on one fragment, so the logic is more like Map-Reduce implementation. This implementation is inspired by https://github.com/tomaszmichalak/vertx-rx-map-reduce.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: core/src/test/java/io/knotx/engine/core/KnotEngineTest.java [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+@ExtendWith(VertxExtension.class)
+public class KnotEngineTest {
+
+  @BeforeEach
+  void setUp() {
+    Assertions.assertTrue(System.getenv().containsKey("vertx.logger-delegate-factory-class-name"));
+  }
+
+  @Test
+  public void execute_whenEventWithNoKnot_expectEventWithUnprocessedStatus(
```
Comment (Skejven) [inline core/src/test/java/io/knotx/engine/core/KnotEngineTest.java]: How about using `@DisplayName("Expect unprocessed status when empty knot flow")`?
And then name method like `executeWithNoKnot` or smth like this?

File: core/src/test/java/io/knotx/engine/core/KnotEngineTest.java [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+              Operation.of("aAddress", "SKIPPED")
+          )));
+    });
+  }
+
+  @Test
+  public void execute_whenEventAndProcessingKnot_expectEventWithProcessedLogEntriesAndSuccessStatus(
+      VertxTestContext testContext, Vertx vertx) throws Throwable {
+    // given
+    createSuccessKnot(vertx, "aAddress", "next");
```
Comment (Skejven) [inline core/src/test/java/io/knotx/engine/core/KnotEngineTest.java]: Is that important here that success knot have `"next"` transition?
Shouldn't be `null`?
How different would be behaviour when transition here is `"next"` or `null`?

File: core/src/test/java/io/knotx/engine/core/KnotEngineTest.java [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+      VertxTestContext testContext, Vertx vertx) throws Throwable {
+    // given
+    createSuccessKnot(vertx, "aAddress", null);
+    createSuccessKnot(vertx, "bAddress", null);
+
+    KnotFlow firstKnotFlow = new KnotFlow("aAddress", Collections.emptyMap());
+    KnotFlow secondKnotFlow = new KnotFlow("bAddress", Collections.emptyMap());
+
+    // when
+    // when
```
Comment (Skejven) [inline core/src/test/java/io/knotx/engine/core/KnotEngineTest.java]: `//then`

File: api/src/main/java/io/knotx/engine/api/KnotFlow.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  }
+
+  public JsonObject toJson() {
+    JsonObject map = new JsonObject();
+    onTransition.forEach((key, value) -> map.put(key, value.toJson()));
+
+    return new JsonObject().put("address", address).put("onTransition", map);
+  }
+
+  private static void fromJson(Iterable<java.util.Map.Entry<String, Object>> json,
```
Comment (Skejven) [inline api/src/main/java/io/knotx/engine/api/KnotFlow.java]: Why `json` is not `JsonObject`?

File: api/src/main/java/io/knotx/engine/api/KnotFlow.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+      KnotFlow knotFlow) {
+    knotFlow.onTransition = Collections.emptyMap();
+    for (java.util.Map.Entry<String, Object> member : json) {
+      switch (member.getKey()) {
```
Comment (Skejven) [inline api/src/main/java/io/knotx/engine/api/KnotFlow.java]: This need to be refactored to more readable code.

File: core/src/main/java/io/knotx/engine/core/impl/KnotEngineImpl.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+import io.vertx.reactivex.core.Vertx;
+import io.vertx.serviceproxy.ServiceException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+class KnotEngineImpl implements KnotEngine {
```
Comment (Skejven) [inline core/src/main/java/io/knotx/engine/core/impl/KnotEngineImpl.java]: `KnotxKnotEngine` or `DefaultKnotEngine`? I'm not happy with `...Impl`

File: core/src/main/java/io/knotx/engine/core/impl/KnotEngineImpl.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+          list.add(item);
+          return list;
+        })
+        .map(this::sortAccordingToIncomigOrder)
+        .map(this::covert);
+  }
+
+  private List<FragmentEventContext> wrapContext(List<FragmentEvent> sourceEvent,
+      ClientRequest clientRequest) {
+    List<FragmentEventContext> result = new ArrayList<>(sourceEvent.size());
```
Comment (Skejven) [inline core/src/main/java/io/knotx/engine/core/impl/KnotEngineImpl.java]: sourceEvent.stream -> map -> collect would be also good here.

File: api/src/main/java/io/knotx/engine/api/KnotFlow.java (hunk 5 above)
Comment (tomaszmichalak) [reply api/src/main/java/io/knotx/engine/api/KnotFlow.java]: added todo

**Pull request merged**